### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Bug: Inconsistent HTML tag validation**

There's an inconsistency in the regex patterns used for HTML tag validation:

- Title uses: `/<[^>]*>/g` (matches `<script>`, rejects ``)
- Description uses: `/<.*>/g` (matches any `<...>` pattern, would reject `` as well)

The title regex is more restrictive and correct for XSS prevention. Both should use the same pattern: `/<[^>]*>/g`

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Bug: Missing validation for ID "0"**

The current regex `/^\d+$/` accepts "0" as a valid ID, but depending on your ID generation strategy (auto-increment starting from 1), ID "0" might not be valid. Consider using `/^[1-9]\d*$/` if you want to exclude 0, or add explicit validation for the minimum ID value.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Potential Issue: Missing error handling**

There's no try-catch block around `todoStore.create()`. If the store throws an exception (e.g., database connection error, validation error), the server will crash with an unhandled exception. Consider wrapping this in a try-catch block to provide graceful error handling.

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **Missing Test Coverage**

The comment indicates missing test coverage for description HTML validation. Consider adding this test:

```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
+    // MISSING TEST: Should test ID "0" edge case - currently no validation for this
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:120]: **Missing Test Coverage for Edge Case**

Consider adding a test for the ID "0" edge case mentioned in the comment:

```typescript
it('should reject ID "0" if not valid in your system', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer greater than 0',
    value: '0'
  });
});
```

Only add this if your ID system doesn't allow 0.

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Overall Review Summary

This PR adds comprehensive input validation to the Todo API, which is a great security and UX improvement. The implementation is well-structured with good separation of concerns, comprehensive test coverage, and clear documentation.

### ✅ **Strengths:**
- **Robust validation system** with detailed, structured error responses
- **Security improvements** with HTML tag prevention for XSS protection  
- **Comprehensive test suite** with 15+ test cases covering various scenarios
- **Consistent error handling** across all endpoints with structured response format
- **Clear documentation** of validation rules and API changes in README
- **Breaking change notification** - good practice for API consumers

### ⚠️ **Issues to Address:**

1. **HTML Validation Inconsistency**: Different regex patterns for title vs description HTML validation
2. **Missing Error Handling**: No try-catch in POST route for todoStore.create()
3. **ID "0" Edge Case**: Current validation allows ID "0" which may not be valid
4. **Missing Test Coverage**: Some test cases mentioned as missing in comments

### 🔧 **Recommendations:**

1. **Fix the HTML regex inconsistency** - use `/<[^>]*>/g` for both title and description
2. **Add error handling** around store operations to prevent server crashes
3. **Clarify ID validation strategy** - decide whether ID "0" should be valid
4. **Add missing test cases** for complete coverage
5. **Consider adding validation middleware to todoStore operations** for additional defense-in-depth

### 🚀 **Future Enhancements:**
The planned future enhancements (rate limiting, internationalization, etc.) show good forward thinking.

**Recommendation: REQUEST_CHANGES** - Please address the HTML validation inconsistency and error handling issues before merging. The other items can be addressed in follow-up PRs if preferred.
